src/gui/sidebar/previews/layer/SidebarPreviewLayers.cpp

```diff
@@ -7,20 +7,82 @@
 #include "SidebarPreviewLayerEntry.h"
 #include "i18n.h"
 
+// Since we have two layer sidebars, we need to make sure we don't
+// double-register callbacks. Since we also need to make sure we only do the
+// proper deallocation stuff on destuction of the last SidebarPreviewLayers
+// object, we use a static counter which we count up in construction and down
+// in destruction.
+static int layerSidebarsActive = 0;
```

Comment:
Thank you for the detailed comment! I don't like having static, mutable variables like this, though...

I think I might like the following better:
 * Create `BaseContextMenu` and `SidebarLayersContextMenu` classes. Give a `std::shared_ptr<SidebarLayersContextMenu>` to both `SidebarPreviewLayers`.
    * I (or anyone else who wants to) could refactor `SidebarPreviewPages` in a follow-up PR
    * This simplifies the code in `SidebarPreviewLayers` and gets rid of the `static` variable.

An issue with my suggestion is that, until we refactor `SidebasrPreviewPages` to use `BaseContextMenu`, there will be a potentially-confusing inconsistency.

Reply:
This is a lot cleaner for sure. I did the static counter mostly because I didn't want to touch too much "other" code, but this was misguided considering how unwieldy this variable would have been. Thanks for the nudge.

Reply:
Switching the context menu for the pages sidebar would look something like this: 

```diff
diff --git a/src/gui/sidebar/previews/page/SidebarPagesContextMenu.cpp b/src/gui/sidebar/previews/page/SidebarPagesContextMenu.cpp
new file mode 100644
index 00000000..b5ff69aa
--- /dev/null
+++ b/src/gui/sidebar/previews/page/SidebarPagesContextMenu.cpp
@@ -0,0 +1,46 @@
+#include "gui/sidebar/previews/page/SidebarPagesContextMenu.h"
+
+#include <memory>
+
+SidebarPagesContextMenu::SidebarPagesContextMenu(GladeGui* gui, SidebarToolbar* toolbar):
+        SidebarBaseContextMenu(gui->get("sidebarPreviewContextMenu")) {
+
+    // Connect the context menu actions
+    const std::map<std::string, SidebarActions> ctxMenuActions = {
+            {"sidebarPreviewDuplicate", SIDEBAR_ACTION_COPY},
+            {"sidebarPreviewDelete", SIDEBAR_ACTION_DELETE},
+            {"sidebarPreviewMoveUp", SIDEBAR_ACTION_MOVE_UP},
+            {"sidebarPreviewMoveDown", SIDEBAR_ACTION_MOVE_DOWN},
+            {"sidebarPreviewNewBefore", SIDEBAR_ACTION_NEW_BEFORE},
+            {"sidebarPreviewNewAfter", SIDEBAR_ACTION_NEW_AFTER},
+    };
+
+    for (const auto& pair: ctxMenuActions) {
+        GtkWidget* const entry = gui->get(pair.first);
+        g_assert(entry != nullptr);
+
+        // Unfortunately, we need a fairly complicated mechanism to keep track
+        // of which action we want to execute.
+        using Data = ContextMenuData;
+        auto userdata = std::make_unique<Data>(Data{toolbar, pair.second});
+
+        const auto callback =
+                G_CALLBACK(+[](GtkMenuItem* item, Data* data) { data->toolbar->runAction(data->actions); });
+        const gulong signalId = g_signal_connect(entry, "activate", callback, userdata.get());
+        g_object_ref(entry);
+        this->contextMenuSignals.emplace_back(entry, signalId, std::move(userdata));
+
+        if (pair.first == "sidebarPreviewMoveDown") {
+            this->contextMenuMoveDown = entry;
+        } else if (pair.first == "sidebarPreviewMoveUp") {
+            this->contextMenuMoveUp = entry;
+        }
+    }
+    g_assert(this->contextMenuMoveDown != nullptr);
+    g_assert(this->contextMenuMoveUp != nullptr);
+}
+
+void SidebarPagesContextMenu::setActionsSensitivity(SidebarActions actions) {
+    gtk_widget_set_sensitive(this->contextMenuMoveUp, actions & SIDEBAR_ACTION_MOVE_UP);
+    gtk_widget_set_sensitive(this->contextMenuMoveDown, actions & SIDEBAR_ACTION_MOVE_DOWN);
+}
diff --git a/src/gui/sidebar/previews/page/SidebarPagesContextMenu.h b/src/gui/sidebar/previews/page/SidebarPagesContextMenu.h
new file mode 100644
index 00000000..31c2eaba
--- /dev/null
+++ b/src/gui/sidebar/previews/page/SidebarPagesContextMenu.h
@@ -0,0 +1,22 @@
+#pragma once
+
+#include <map>
+#include <memory>
+#include <vector>
+
+#include <gtk/gtk.h>
+
+#include "gui/sidebar/previews/base/SidebarBaseContextMenu.h"
+#include "gui/sidebar/previews/base/SidebarToolbar.h"
+
+
+class SidebarPagesContextMenu: public SidebarBaseContextMenu {
+public:
+    SidebarPagesContextMenu(GladeGui* gui, SidebarToolbar* toolbar);
+
+    void setActionsSensitivity(SidebarActions actions) override;
+
+private:
+    GtkWidget* contextMenuMoveUp = nullptr;
+    GtkWidget* contextMenuMoveDown = nullptr;
+};
diff --git a/src/gui/sidebar/previews/page/SidebarPreviewPages.cpp b/src/gui/sidebar/previews/page/SidebarPreviewPages.cpp
index 5b9126d4..61fc4700 100644
--- a/src/gui/sidebar/previews/page/SidebarPreviewPages.cpp
+++ b/src/gui/sidebar/previews/page/SidebarPreviewPages.cpp
@@ -13,53 +13,8 @@
 
 SidebarPreviewPages::SidebarPreviewPages(Control* control, GladeGui* gui, SidebarToolbar* toolbar):
         SidebarPreviewBase(control, gui, toolbar),
-        contextMenu(gui->get("sidebarPreviewContextMenu")),
-        iconNameHelper(control->getSettings()) {
-    // Connect the context menu actions
-    const std::map<std::string, SidebarActions> ctxMenuActions = {
-            {"sidebarPreviewDuplicate", SIDEBAR_ACTION_COPY},
-            {"sidebarPreviewDelete", SIDEBAR_ACTION_DELETE},
-            {"sidebarPreviewMoveUp", SIDEBAR_ACTION_MOVE_UP},
-            {"sidebarPreviewMoveDown", SIDEBAR_ACTION_MOVE_DOWN},
-            {"sidebarPreviewNewBefore", SIDEBAR_ACTION_NEW_BEFORE},
-            {"sidebarPreviewNewAfter", SIDEBAR_ACTION_NEW_AFTER},
-    };
-
-    for (const auto& pair: ctxMenuActions) {
-        GtkWidget* const entry = gui->get(pair.first);
-        g_assert(entry != nullptr);
-
-        // Unfortunately, we need a fairly complicated mechanism to keep track
-        // of which action we want to execute.
-        using Data = SidebarPreviewPages::ContextMenuData;
-        auto userdata = std::make_unique<Data>(Data{this->toolbar, pair.second});
-
-        const auto callback =
-                G_CALLBACK(+[](GtkMenuItem* item, Data* data) { data->toolbar->runAction(data->actions); });
-        const gulong signalId = g_signal_connect(entry, "activate", callback, userdata.get());
-        g_object_ref(entry);
-        this->contextMenuSignals.emplace_back(entry, signalId, std::move(userdata));
-
-        if (pair.first == "sidebarPreviewMoveDown") {
-            this->contextMenuMoveDown = entry;
-        } else if (pair.first == "sidebarPreviewMoveUp") {
-            this->contextMenuMoveUp = entry;
-        }
-    }
-    g_assert(this->contextMenuMoveDown != nullptr);
-    g_assert(this->contextMenuMoveUp != nullptr);
-}
-
-SidebarPreviewPages::~SidebarPreviewPages() {
-    for (const auto& signalTuple: this->contextMenuSignals) {
-        GtkWidget* const widget = std::get<0>(signalTuple);
-        const guint handlerId = std::get<1>(signalTuple);
-        if (g_signal_handler_is_connected(widget, handlerId)) {
-            g_signal_handler_disconnect(widget, handlerId);
-        }
-        g_object_unref(widget);
-    }
-}
+        contextMenu(std::make_shared<SidebarPagesContextMenu>(gui, toolbar)),
+        iconNameHelper(control->getSettings()) {}
 
 void SidebarPreviewPages::enableSidebar() {
     SidebarPreviewBase::enableSidebar();
@@ -299,11 +254,9 @@ void SidebarPreviewPages::pageSelected(size_t page) {
         this->toolbar->setHidden(false);
         this->toolbar->setButtonEnabled(static_cast<SidebarActions>(actions));
 
-        gtk_widget_set_sensitive(this->contextMenuMoveUp, actions & SIDEBAR_ACTION_MOVE_UP);
-        gtk_widget_set_sensitive(this->contextMenuMoveDown, actions & SIDEBAR_ACTION_MOVE_DOWN);
+        this->contextMenu->setActionsSensitivity(SidebarActions(actions));
     }
 }
 
-void SidebarPreviewPages::openPreviewContextMenu() {
-    gtk_menu_popup(GTK_MENU(this->contextMenu), nullptr, nullptr, nullptr, nullptr, 3, gtk_get_current_event_time());
-}
+// TODO: does this get used?!
+void SidebarPreviewPages::openPreviewContextMenu() { this->contextMenu->open(); }
diff --git a/src/gui/sidebar/previews/page/SidebarPreviewPages.h b/src/gui/sidebar/previews/page/SidebarPreviewPages.h
index d5c36241..eb19689b 100644
--- a/src/gui/sidebar/previews/page/SidebarPreviewPages.h
+++ b/src/gui/sidebar/previews/page/SidebarPreviewPages.h
@@ -17,13 +17,12 @@
 #include <vector>
 
 #include "gui/sidebar/previews/base/SidebarPreviewBase.h"
+#include "gui/sidebar/previews/page/SidebarPagesContextMenu.h"
 #include "util/IconNameHelper.h"
 
-
 class SidebarPreviewPages: public SidebarPreviewBase {
 public:
     SidebarPreviewPages(Control* control, GladeGui* gui, SidebarToolbar* toolbar);
-    virtual ~SidebarPreviewPages();
 
 public:
     /**
@@ -69,28 +68,8 @@ private:
      */
     void unselectPage();
 
-    /**
-     * The context menu to display when a page is right-clicked.
-     */
-    GtkWidget* const contextMenu = nullptr;
-
-    GtkWidget* contextMenuMoveUp = nullptr;
-    GtkWidget* contextMenuMoveDown = nullptr;
-
-    /**
-     * The data passed to the menu item callbacks.
-     */
-    struct ContextMenuData {
-        SidebarToolbar* toolbar;
-        SidebarActions actions;
-    };
-
+    std::shared_ptr<SidebarPagesContextMenu> contextMenu;
 
-    /**
-     * The signals connected to the context menu items. This must be kept track
-     * of so the data can be deallocated safely.
-     */
-    std::vector<std::tuple<GtkWidget*, gulong, std::unique_ptr<ContextMenuData>>> contextMenuSignals;
 
 private:
     IconNameHelper iconNameHelper;
```